<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>A more ambitious real-world example &mdash; ruffus v1.04 documentation</title>
    <link rel="stylesheet" href="_static/ruffus.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.04',
        COLLAPSE_MODINDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="ruffus v1.04 documentation" href="index.html" />
    <link rel="next" title="How to share data across jobs" href="sharing_data_across_jobs_example.html" />
    <link rel="prev" title="A real-world example" href="intermediate_example.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="modindex.html" title="Global Module Index"
             accesskey="M">modules</a> |</li>
        <li class="right" >
          <a href="sharing_data_across_jobs_example.html" title="How to share data across jobs"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="intermediate_example.html" title="A real-world example"
             accesskey="P">previous</a> |</li>
        Ruffus v. 1.04          :   
        <li><a href="index.html">Home</a> | </li>
        <li><a href="contents.html">Contents</a> | </li>
        <li><a href="cheatsheet.html">Cheat sheet</a> | </li>
        <li><a href="Tutorial.html">Full tutorial</a>  | </li>
        <li><a href="FAQ.html">FAQ</a>  | </li>
        <li><a href="Todo.html">Future plans</a>  | </li>
        <li><a href="search.html">Search</a> &raquo;</li>
 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="a-more-ambitious-real-world-example">
<span id="complicated-pipelines"></span><h1>A more ambitious real-world example<a class="headerlink" href="#a-more-ambitious-real-world-example" title="Permalink to this headline">¶</a></h1>
<div class="section" id="python-code">
<span id="index-261"></span><h2>Python Code<a class="headerlink" href="#python-code" title="Permalink to this headline">¶</a></h2>
<p>The full code is available <a class="reference external" href="complicated_example_code.html#code-for-complicated-pipeline-example"><em>here</em></a>.</p>
</div>
<div class="section" id="understanding-newly-discovered-genes">
<h2>Understanding newly discovered genes<a class="headerlink" href="#understanding-newly-discovered-genes" title="Permalink to this headline">¶</a></h2>
<p>This is the sort of problem that motivated the design of the <em>ruffus</em> module.</p>
<div class="section" id="background">
<h3>Background:<a class="headerlink" href="#background" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>We have been asked to understand the function of some newly discovered genes:</p>
<blockquote>
<p>With a bit of luck, someone, somewhere has previous investigated the corresponding genes
in e.g. the fly or octopus. This would save us a lot of work.</p>
<p>So, first we have to find the <em>corresponding</em> genes
(<a class="reference external" href="http://en.wikipedia.org/wiki/Homology_(biology)">homologs</a>) in different species.</p>
</blockquote>
<p>Comparisons within each <em>gene family</em> of corresponding genes can be very useful too:</p>
<blockquote>
We can see how similar parts of each gene have been evolving.
Crucial active sites in enzymes can stay unchanged for hundreds
of millions of years. Other parts can change very rapidly because we
are in an arms race with parasites, viruses and the like...</blockquote>
</blockquote>
</div>
<div class="section" id="aim">
<h3>Aim:<a class="headerlink" href="#aim" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>We need to:</p>
<ul class="simple">
<li>Find all genes related to our unknown genes in different species.</li>
<li>create <em>family trees</em> for each sets of related genes</li>
<li>Line up corresponding parts of each gene</li>
<li>Analyse their evolution.</li>
</ul>
<p>Luckily, there are standard tools to do all this!!</p>
<p>These are computationally intensive but <a class="reference external" href="http://en.wikipedia.org/wiki/Embarrassingly_parallel">embarassingly parallel</a>
tasks. We just have to break them up into digestible pieces running in
parallel on our cheap computational clusters.</p>
<p>This is where <em>ruffus</em> can come in.</p>
</blockquote>
</div>
<div class="section" id="conceptual-flow-chart">
<h3>Conceptual Flow chart:<a class="headerlink" href="#conceptual-flow-chart" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>Our pipeline has 5 easy conceptual steps:</p>
<ol class="arabic simple">
<li>Compare all the unknown gene sets with all genes from each species</li>
<li>Put together gene families from relationships involve all the species</li>
<li>Analyse the evolution of each gene family</li>
<li>Summarise per (previously) unknown gene set</li>
<li>Produce overall summary</li>
</ol>
<img alt="_images/complex_conceptual.jpg" src="_images/complex_conceptual.jpg" />
<p>This is a simple list of sequential tasks which are conceptually easy (for genome
scientists!) to grasp, and should be straightforward to code using <em>ruffus</em>.</p>
</blockquote>
</div>
<div class="section" id="flow-chart-of-file-dependencies">
<h3>Flow chart of file dependencies:<a class="headerlink" href="#flow-chart-of-file-dependencies" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>There are two reasons why this pipeline is more complicated than it seems:</p>
<ul>
<li><p class="first">Large files have to be broken up into small chunks
and parcelled out to run in parallel.</p>
<p>The results then have to be recombined into a single file.</p>
</li>
<li><p class="first">An all vs. all comparison is involved</p>
</li>
</ul>
<p>This complexity is obvious in even this simplified flowchart:</p>
<img alt="_images/complex_file_dag.jpg" src="_images/complex_file_dag.jpg" />
<p>These operations are <em>particularly</em> horrible to describe using
<tt class="docutils literal"><span class="pre">gmake</span></tt> or <tt class="docutils literal"><span class="pre">scons</span></tt> which rely on file dependency trees:</p>
<blockquote>
Each line in step (3) is an operation between two different files.</blockquote>
<p><tt class="docutils literal"><span class="pre">gmake</span></tt> or <tt class="docutils literal"><span class="pre">scons</span></tt> much prefer that we know up front how many files are involved
in the pipeline.</p>
<p>In this and similar cases, we <em>can&#8217;t</em> know in advance of running the pipeline how
many small chunks will be produced. It depends on the size of gene sets and
the size of the calculated gene families.</p>
<p>To code this in <tt class="docutils literal"><span class="pre">gmake</span></tt> or <tt class="docutils literal"><span class="pre">scons</span></tt> require either recursive calls, splitting up
the pipeline, or horrible hacks to the internals of each tool.</p>
</blockquote>
</div>
</div>
<div class="section" id="ruffus-flow-chart">
<h2><em>Ruffus</em> Flow chart<a class="headerlink" href="#ruffus-flow-chart" title="Permalink to this headline">¶</a></h2>
<blockquote>
<p>Our pipeline flow chart is built around these 9 steps:</p>
<ol class="arabic simple">
<li>Split up the unknown gene sets into smaller chunks</li>
<li>Likewise split up the sets of known gene sets from each species</li>
<li>Run each unknown chunk against <em>all</em> the known chunks (all-vs-all comparisons)</li>
<li>Combine chunk comparisons into gene families</li>
<li>Split up gene families into bite-sizes</li>
<li>Analyse evolutionary parameters</li>
<li>Recombine results per unknown gene set</li>
<li>Summary per unknown gene set</li>
<li>Overall summary</li>
</ol>
<img alt="_images/complex_ruffus.jpg" src="_images/complex_ruffus.jpg" />
<p>Extra tasks 1, 2, 4, 5 and 7 involve splitting up the work for
parallel jobs, or re-combining them afterwards.</p>
<p>Let us go through the more involved of these steps.</p>
<p>The full code is available <a class="reference external" href="complicated_example_code.html#code-for-complicated-pipeline-example"><em>here</em></a>.</p>
</blockquote>
<div class="section" id="split-up-unknown-gene-sets">
<h3>1. Split up unknown gene sets<a class="headerlink" href="#split-up-unknown-gene-sets" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>We first split up the unknown gene sets into smaller chunks for
parallel processing.</p>
<p>We make sure the working directory is created using:</p>
<div class="highlight-python"><pre>@follows(mkdir(...))</pre>
</div>
<p>The files can then be created  as follows:</p>
<div class="highlight-python"><pre>unknown_genes/XXX.genes.fa
    -&gt;  XXX/split_gene_sets.completed
        XXX/NNN.fa</pre>
</div>
<p>where <tt class="docutils literal"><span class="pre">XXX</span></tt> is the name of the unknown gene set, and <tt class="docutils literal"><span class="pre">NNN.fa</span></tt> represent successively
numbered files, e.g. <tt class="docutils literal"><span class="pre">1.fa</span></tt>, <tt class="docutils literal"><span class="pre">2.fa</span></tt> and so on.</p>
<p>Since we don&#8217;t known how many file chunks will be produced, each job creates a
single flag file (<tt class="docutils literal"><span class="pre">split_gene_sets.completed</span></tt>) to indicate it has run successfully.</p>
<p>The python code will look this this:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">unknown_gene_sets</span> <span class="o">=</span> <span class="n">data_dir</span> <span class="o">+</span> <span class="s">&quot;/unknown_genes/*.genes.fa&quot;</span>

<span class="nd">@follows</span><span class="p">(</span><span class="n">mkdir</span><span class="p">(</span><span class="n">working_dir</span><span class="p">))</span>

<span class="nd">@files_re</span><span class="p">(</span><span class="n">unknown_gene_sets</span><span class="p">,</span>
          <span class="s">r&quot;(.*/)(.*)(.genes.fa)&quot;</span><span class="p">,</span>                         <span class="c"># regular expression</span>
          <span class="s">r&quot;\1\2\3&quot;</span><span class="p">,</span>                                       <span class="c"># starting_gene_set</span>
          <span class="n">working_dir</span> <span class="o">+</span> <span class="s">r&quot;/\2/split_gene_sets.completed&quot;</span><span class="p">,</span>  <span class="c"># job_completion_flag</span>
          <span class="n">working_dir</span> <span class="o">+</span> <span class="s">r&quot;/\2&quot;</span><span class="p">)</span>                            <span class="c"># output_dir</span>
<span class="k">def</span> <span class="nf">split_unknown_gene_set</span><span class="p">(</span> <span class="n">starting_gene_set</span><span class="p">,</span>
                            <span class="n">job_completion_flag</span><span class="p">,</span>
                            <span class="n">split_output_dir</span><span class="p">):</span>
    <span class="n">make_directory_for_chunks</span><span class="p">()</span>
    <span class="n">split_unknown_genes_sets_into_chunks</span><span class="p">()</span>
    <span class="n">touch_completion_flag_file</span><span class="p">()</span>
</pre></div>
</div>
</blockquote>
</div>
<div class="section" id="split-up-known-gene-sets-for-each-species">
<h3>2. Split up known gene sets for each species<a class="headerlink" href="#split-up-known-gene-sets-for-each-species" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>Do likewise with the known gene sets from each of the species we are comparing against.</p>
<p>The files are created  as follows:</p>
<div class="highlight-python"><pre>all_genes_in_each_species/YYY.genes.fa
    -&gt; species_YYY/split_gene_sets.completed
    -&gt; species_YYY/MMM.fa</pre>
</div>
<p>where <tt class="docutils literal"><span class="pre">YYY</span></tt> is a species name e.g. <tt class="docutils literal"><span class="pre">fly</span></tt> or <tt class="docutils literal"><span class="pre">octopus</span></tt>, and <tt class="docutils literal"><span class="pre">MMM.fa</span></tt> represent
successively numbered files</p>
</blockquote>
</div>
<div class="section" id="all-vs-all-comparisons">
<h3>3. All vs. all comparisons<a class="headerlink" href="#all-vs-all-comparisons" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>Now we need to compare all the chunks in each species against all the chunks in
each unknown gene set:</p>
<div class="highlight-python"><pre>for x in XXX:
    for n in NNN:
        for y in YYY:
            for m in MMM:
                    x/n.fa vs. species_y/m.fa
                        -&gt; compare/x/y.n.m.comparison_res
                        -&gt; compare/x/y.n.m.complete</pre>
</div>
<dl class="docutils">
<dt>where</dt>
<dd><p class="first"><tt class="docutils literal"><span class="pre">XXX</span></tt> is the name of the unknown gene set,</p>
<p><tt class="docutils literal"><span class="pre">YYY</span></tt> is the species name e.g. <tt class="docutils literal"><span class="pre">fly</span></tt> or <tt class="docutils literal"><span class="pre">octopus</span></tt>,</p>
<p class="last"><tt class="docutils literal"><span class="pre">NNN</span></tt> and <tt class="docutils literal"><span class="pre">MMM</span></tt> represent successively numbered files</p>
</dd>
</dl>
<p>We also use a completion flag file (<tt class="docutils literal"><span class="pre">compare/x/y.n.m.complete</span></tt>) because if
no related genes (<a class="reference external" href="http://en.wikipedia.org/wiki/Homology_(biology)">homologs</a>) are found
between the chunk of unknown genes and the chunk of species genes, no output
will be produced. The completion flag is the only way we will know that the
job completed properly.</p>
<p>These all vs all comparisons are easy to set up with a custom python function.
The most important part of the code involves the two sets of nested for loops:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">generate_all_vs_all_params</span> <span class="p">():</span>

    <span class="n">species_names</span> <span class="o">=</span> <span class="n">get_species_names</span><span class="p">()</span>
    <span class="n">gene_set_names</span> <span class="o">=</span> <span class="n">get_unknown_gene_set_names</span><span class="p">()</span>

    <span class="c">#</span>
    <span class="c">#   Nested for loops (1):</span>
    <span class="c">#</span>
    <span class="c">#       each species vs each unknown gene set</span>
    <span class="c">#</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">gene_set_names</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">species_names</span><span class="p">:</span>

            <span class="n">m_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s">/</span><span class="si">%s</span><span class="s">/*.fa&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">w_dir</span><span class="p">,</span> <span class="n">x</span><span class="p">))</span>
            <span class="n">n_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s">&quot;</span><span class="si">%s</span><span class="s">/</span><span class="si">%s</span><span class="s">/*.fa&quot;</span> <span class="o">%</span> <span class="p">(</span><span class="n">w_dir</span><span class="p">,</span> <span class="n">y</span><span class="p">))</span>

            <span class="c">#</span>
            <span class="c">#   Nested for loops (2):</span>
            <span class="c">#</span>
            <span class="c">#       each species chunk vs for each unknown chunk</span>
            <span class="c">#</span>
            <span class="k">for</span> <span class="n">m_file</span> <span class="ow">in</span> <span class="n">m_files</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">n_file</span> <span class="ow">in</span> <span class="n">n_files</span><span class="p">:</span>
                    <span class="n">input_files</span> <span class="o">=</span> <span class="p">[</span><span class="n">m_file</span><span class="p">,</span> <span class="n">n_file</span><span class="p">]</span>

                    <span class="k">yield</span> <span class="n">input_files</span> <span class="c"># more code removed ...</span>
</pre></div>
</div>
</blockquote>
</div>
<div class="section" id="combine-chunk-comparisons-into-gene-families">
<h3>4. Combine chunk comparisons into gene families<a class="headerlink" href="#combine-chunk-comparisons-into-gene-families" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="recombine-results-per-unknown-gene-set">
<h3>7. Recombine results per unknown gene set<a class="headerlink" href="#recombine-results-per-unknown-gene-set" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>Steps (4) and (7) involve gathering many files and merging them together in one
output file. The best way to do this again involves generating the job parameters
on the fly.</p>
<p>This turns out to be a really easy function to write. We just need to specify
all the input files from a glob:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="k">def</span> <span class="nf">generate_params_for_making_gene_families</span> <span class="p">():</span>
    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">get_unknown_gene_set_names</span><span class="p">():</span>
        <span class="n">results_files</span> <span class="o">=</span> <span class="n">glob</span><span class="o">.</span><span class="n">glob</span><span class="p">(</span><span class="s">&quot;compare/</span><span class="si">%s</span><span class="s">/*.comparison_res&quot;</span> <span class="o">%</span> <span class="n">x</span><span class="p">)</span>
        <span class="n">family_file</span> <span class="o">=</span> <span class="s">&quot;multiple_alignment/</span><span class="si">%s</span><span class="s">/gene.families&quot;</span> <span class="o">%</span> <span class="n">x</span>
        <span class="k">yield</span> <span class="n">results_files</span><span class="p">,</span> <span class="n">family_file</span>
</pre></div>
</div>
</blockquote>
</div>
<div class="section" id="split-up-gene-families-lists">
<h3>5. Split up gene families lists<a class="headerlink" href="#split-up-gene-families-lists" title="Permalink to this headline">¶</a></h3>
<blockquote>
The uses the same approach as steps (1) and (2)</blockquote>
</div>
<div class="section" id="analyse-evolutionary-parameters">
<h3>6. Analyse evolutionary parameters<a class="headerlink" href="#analyse-evolutionary-parameters" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="summary-per-unknown-gene-set">
<h3>8. Summary per unknown gene set<a class="headerlink" href="#summary-per-unknown-gene-set" title="Permalink to this headline">¶</a></h3>
<blockquote>
<p>Steps (6) and (8) turned out to be the easiest parts of the entire pipeline:
Generating files of one type from another with two suffices involves the simplest
of regular expressions:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="nd">@files_re</span><span class="p">(</span><span class="s">&quot;multiple_alignment/*/*.aln&quot;</span><span class="p">,</span> <span class="s">r&quot;aln$&quot;</span><span class="p">,</span> <span class="s">r&quot;evo_res&quot;</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">evolution_analysis</span><span class="p">(</span> <span class="n">family_file</span><span class="p">,</span> <span class="n">result_file_name</span><span class="p">):</span>
    <span class="s">&quot; ... &quot;</span>
</pre></div>
</div>
</blockquote>
</div>
<div class="section" id="overall-summary">
<h3>9. Overall summary<a class="headerlink" href="#overall-summary" title="Permalink to this headline">¶</a></h3>
<blockquote>
We simple supplied a list of the separate summary files to the <tt class="docutils literal"><span class="pre">&#64;files</span></tt> decorator
to generate the final conclusions of this pipline. Couldn&#8217;t be simpler.</blockquote>
</div>
<div class="section" id="conclusions">
<h3>Conclusions<a class="headerlink" href="#conclusions" title="Permalink to this headline">¶</a></h3>
<p>This may seem a long-winded example but without
<em>ruffus</em> to help abstract away some of the complexity, the
code would have been quite unmanageable.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
            <h3><a href="contents.html">Table Of Contents</a></h3>
            <ul>
<li><a class="reference external" href="">A more ambitious real-world example</a><ul>
<li><a class="reference external" href="#python-code">Python Code</a></li>
<li><a class="reference external" href="#understanding-newly-discovered-genes">Understanding newly discovered genes</a><ul>
<li><a class="reference external" href="#background">Background:</a></li>
<li><a class="reference external" href="#aim">Aim:</a></li>
<li><a class="reference external" href="#conceptual-flow-chart">Conceptual Flow chart:</a></li>
<li><a class="reference external" href="#flow-chart-of-file-dependencies">Flow chart of file dependencies:</a></li>
</ul>
</li>
<li><a class="reference external" href="#ruffus-flow-chart"><em>Ruffus</em> Flow chart</a><ul>
<li><a class="reference external" href="#split-up-unknown-gene-sets">1. Split up unknown gene sets</a></li>
<li><a class="reference external" href="#split-up-known-gene-sets-for-each-species">2. Split up known gene sets for each species</a></li>
<li><a class="reference external" href="#all-vs-all-comparisons">3. All vs. all comparisons</a></li>
<li><a class="reference external" href="#combine-chunk-comparisons-into-gene-families">4. Combine chunk comparisons into gene families</a></li>
<li><a class="reference external" href="#recombine-results-per-unknown-gene-set">7. Recombine results per unknown gene set</a></li>
<li><a class="reference external" href="#split-up-gene-families-lists">5. Split up gene families lists</a></li>
<li><a class="reference external" href="#analyse-evolutionary-parameters">6. Analyse evolutionary parameters</a></li>
<li><a class="reference external" href="#summary-per-unknown-gene-set">8. Summary per unknown gene set</a></li>
<li><a class="reference external" href="#overall-summary">9. Overall summary</a></li>
<li><a class="reference external" href="#conclusions">Conclusions</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            <h4>Previous topic</h4>
            <p class="topless"><a href="intermediate_example.html"
                                  title="previous chapter">A real-world example</a></p>
            <h4>Next topic</h4>
            <p class="topless"><a href="sharing_data_across_jobs_example.html"
                                  title="next chapter">How to share data across jobs</a></p>
            <h3>This Page</h3>
            <ul class="this-page-menu">
              <li><a href="_sources/complicated_example.txt"
                     rel="nofollow">Show Source</a></li>
            </ul>
          <div id="searchbox" style="display: none">
            <h3>Quick search</h3>
              <form class="search" action="search.html" method="get">
                <input type="text" name="q" size="18" />
                <input type="submit" value="Go" />
                <input type="hidden" name="check_keywords" value="yes" />
                <input type="hidden" name="area" value="default" />
              </form>
              <p class="searchtip" style="font-size: 90%">
              Enter search terms or a module, class or function name.
              </p>
          </div>
          <script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="modindex.html" title="Global Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="sharing_data_across_jobs_example.html" title="How to share data across jobs"
             >next</a> |</li>
        <li class="right" >
          <a href="intermediate_example.html" title="A real-world example"
             >previous</a> |</li>
        Ruffus v. 1.04          :   
        <li><a href="index.html">Home</a> | </li>
        <li><a href="contents.html">Contents</a> | </li>
        <li><a href="cheatsheet.html">Cheat sheet</a> | </li>
        <li><a href="Tutorial.html">Full tutorial</a>  | </li>
        <li><a href="FAQ.html">FAQ</a>  | </li>
        <li><a href="Todo.html">Future plans</a>  | </li>
        <li><a href="search.html">Search</a> &raquo;</li>
 
      </ul>
    </div>
    <div class="footer">
      &copy; Copyright 2009, Leo Goodstadt.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 0.6.1.
    </div>
  </body>
</html>